642 feature request steel RHS and SHS | change of analytical to meshed analysis

GerjanDorgelo opened the pull request:
## Description

adds the RHS and SHS shapes, using the quarter corner file and some general steel changes

All changes:
- some minor documentation changes in I and CHS profiles (e.g. the resolution parameter will get deprecationed soon)
- removed analytical solutions of building blocks
- made corner block for RHS (they have inconsistent width along the corner, otherwise annular sector wouldve worked)
- made RHS from the building blocks
- extended the steel shape example
- split the H-section-flanges into two.

RHS shape as follows:
<img width="979" height="805" alt="afbeelding" src="https://github.com/user-attachments/assets/a1b35ebf-0b37-4e6c-bdcd-7321062d76d9" />

But also allows for:
<img width="1041" height="715" alt="afbeelding" src="https://github.com/user-attachments/assets/cb22bf31-98da-4c64-8868-db910ca1b211" />

Fixes #642

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## Checklist:

- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] New and existing unit tests pass locally with my changes

bro-wi commented on blueprints/structural_sections/cross_section_rhs_corner.py:
```diff
@@ -0,0 +1,129 @@
+    def __post_init__(self) -> None:
+        """Validate input parameters after initialization."""
+        if self.thickness_vertical <= 0:
+            raise ValueError(f"Thickness vertical must be positive, got {self.thickness_vertical}")
+        if self.thickness_horizontal <= 0:
+            raise ValueError(f"Thickness horizontal must be positive, got {self.thickness_horizontal}")
+        if self.inner_radius <= 0:
+            raise ValueError(f"Inner radius must be positive, got {self.inner_radius}")
+        if self.outer_radius <= 0:
+            raise ValueError(f"Outer radius must be positive, got {self.outer_radius}")
```
make use of the custom errors we have.

GerjanDorgelo replied:
rhs corner file has been deleted and is now linked to the spandrel

bro-wi commented on blueprints/structural_sections/cross_section_rhs_corner.py:
```diff
@@ -0,0 +1,129 @@
+    outer_radius: MM
+    x: MM = 0
+    y: MM = 0
+    mirrored_horizontally: bool = False
+    mirrored_vertically: bool = False
```
isn't it more intuitive to input a quadrant instead of two bools?
| 2  |  1 |
|---|----|
|  3  |  4 |

bro-wi commented on blueprints/structural_sections/cross_section_rhs_corner.py:
```diff
@@ -0,0 +1,129 @@
+        n = 16
+
+        # Outer arc (from vertical to horizontal)
+        outer_arc = [
+            (
+                self.x + self.width_rectangle - self.outer_radius + self.outer_radius * math.cos(math.radians(0 + i * 90 / (n - 1))),
+                self.y + self.height_rectangle - self.outer_radius + self.outer_radius * math.sin(math.radians(0 + i * 90 / (n - 1))),
+            )
+            for i in range(n)
```
also here, make use of numpy, it is much faster than a regular python loop.

bro-wi commented on blueprints/structural_sections/cross_section_rhs_corner.py:
```diff
@@ -0,0 +1,129 @@
+        ]
+
+        # Inner arc (from horizontal to vertical, reversed)
+        inner_arc = [
+            (
+                self.x + self.inner_radius * math.cos(math.radians(0 + i * 90 / (n - 1))),
+                self.y + self.inner_radius * math.sin(math.radians(0 + i * 90 / (n - 1))),
+            )
+            for i in range(n)
+        ][::-1]
```
numpy

bro-wi commented on blueprints/structural_sections/cross_section_rhs_corner.py:
```diff
@@ -0,0 +1,129 @@
+
+        points = [lr, *outer_arc, ul, *inner_arc]
+        # Remove consecutive duplicate points
+        points = [pt for i, pt in enumerate(points) if i == 0 or pt != points[i - 1]]
```
use itertools pairwise:

```
points = [points[0]] + [b for a, b in pairwise(points) if a != b]
```
or even better (numpy again) 
```
diffs = np.any(np.diff(points, axis=0) != 0, axis=1)
mask = np.concatenate([[True], diffs])  # Always keep the first point
points = points[mask]
```

bro-wi commented on blueprints/structural_sections/cross_section_rhs_corner.py:
```diff
@@ -0,0 +1,129 @@
+        # Remove consecutive duplicate points
+        points = [pt for i, pt in enumerate(points) if i == 0 or pt != points[i - 1]]
+
+        if self.mirrored_horizontally:
+            points = [(2 * self.x - x, y) for x, y in points]
+        if self.mirrored_vertically:
+            points = [(x, 2 * self.y - y) for x, y in points]
```
numpy (you are probably getting sick of it:P ), and I would even consider doing that with a matrix multiplication using a mirroring matrix

bro-wi commented on blueprints/structural_sections/cross_section_rhs_corner.py:
```diff
@@ -0,0 +1,129 @@
+        Parameters
+        ----------
+        mesh_size : MM | None
+            Maximum mesh element area to be used within the Geometry-object finite-element mesh. If not provided, a default value will be used.
```
and what is the default value ? be explicit to a user.

bro-wi commented on blueprints/structural_sections/steel/steel_cross_sections/rhs_profile.py:
```diff
@@ -0,0 +1,313 @@
+
+        name = profile.alias
+        if corrosion_inside or corrosion_outside:
+            name += (
+                f" (corrosion {'' if not corrosion_inside else f'in: {corrosion_inside} mm'}"
+                f"{', ' if corrosion_inside and corrosion_outside else ''}"
+                f"{'' if not corrosion_outside else f'out: {corrosion_outside} mm'})"
+            )
```
```suggestion
           name += " (corrosion " + ", ".join(
                part for part in (
                    f"in: {corrosion_inside} mm" if corrosion_inside else "",
                    f"out: {corrosion_outside} mm" if corrosion_outside else ""
                ) if part
            ) + ")"
```

GerjanDorgelo replied:
im not sure if this leaves better readibility. imma keep it as it was i think.

bro-wi commented on tests/structural_sections/steel/steel_cross_sections/test_rhs_profile.py:
```diff
@@ -0,0 +1,71 @@
+            profile=RHS.RHS400x200_16,
+            steel_material=SteelMaterial(SteelStrengthClass.S355),
+            corrosion_outside=1,  # mm
+            corrosion_inside=1,  # mm
```
check with different values inside/outside otherwise you won't know if there are placed correctly

bro-wi reviewed (approved):
Nice addition, this would also be a very big addition to blueprints that would open a lot more possibilities! Can't wait too use this in production!

To my taste this PR is way too big. It could have been split up like this:
- Clean up
- PR per css type
- Update documentation

That  makes both checking and iterating much easier and cleaner. The risk now is that this PR will stay on the shelves longer than it should be, while we can release it gradually.

Furthermore i would really invest in using more numpy logic. it already is a dependency and it is way,way faster than any python loop you can write. So we might aswell use it!

GerjanDorgelo commented on blueprints/structural_sections/cross_section_annular_sector.py:
```diff
@@ -128,187 +127,6 @@ def polygon(self) -> Polygon:
         result = outer_ring.difference(inner_ring).intersection(sector)
         return Polygon(result)  # type: ignore[arg-type]
 
-    @property
```
yes. having it all calculated analitically sounded nice, but appeared to be more a nightmare than it offered. So all analytical solutions are deleted.

GerjanDorgelo commented on blueprints/structural_sections/cross_section_quarter_circular_spandrel.py:
```diff
@@ -49,172 +66,37 @@ def polygon(self) -> Polygon:
-            (self.x + self.radius - self.radius * math.cos(math.pi / 2 * i / 25), self.y + self.radius - self.radius * math.sin(math.pi / 2 * i / 25))
+            (self.radius - self.radius * math.cos(math.pi / 2 * i / 25), self.radius - self.radius * math.sin(math.pi / 2 * i / 25))
             for i in range(26)
         ]
```
i cannot make the 26 user input while keeping the function a @ property. i propose leaving it at current value and expending upon it later.

bro-wi replied:
Indeed that wouldn't work for a property. As alternative you could consider implementing it as an optional attribute. An even better/nicer implementation is implementing it as a ClassVar. But I agree, leave it as is for now.

GerjanDorgelo commented:
@bro-wi hi, thanks for the review. 

I implemented some numpy (lets not kid around here, i implemented a lot of numpy), and implemented most of your suggestions. there were two that bugged me: one regarding user input of approximation of quarter circle: that turned out difficult while keeping it as a property. other thing was a suggestion regarding text combination, i tried your way but my code ended up more messy in my opinion.. so i reverted that change. see logs above.

final note: for my taste the PR also grew a bit large. objective was inititally to only add the rhs part, but the analytical values became a pain.

GerjanDorgelo commented:
@bro-wi i re-requested the review. i see you already gave a ✔️, but lets do another round and then you can leave more comments or press the nice green Merge button.

bro-wi reviewed (approved):
Almost there!

Just some small remarks to make things more clear.

egarciamendez commented:
@GerjanDorgelo @bro-wi 
😢 
<img width="1543" height="814" alt="afbeelding" src="https://github.com/user-attachments/assets/2783bf65-0b39-4434-b1b1-79e24f6a8865" />
